File: android-lib/src/androidTest/java/com/kinvey/androidTest/TestManager.java
package com.kinvey.androidTest;

import android.os.Message;

import com.google.api.client.json.GenericJson;
import com.kinvey.android.Client;
import com.kinvey.android.model.User;
import com.kinvey.android.store.DataStore;
import com.kinvey.android.store.UserStore;
import com.kinvey.androidTest.callback.CustomKinveyClientCallback;
import com.kinvey.androidTest.callback.CustomKinveyListCallback;
import com.kinvey.androidTest.callback.CustomKinveyReadCallback;
import com.kinvey.androidTest.callback.CustomKinveyLiveServiceCallback;
import com.kinvey.androidTest.callback.CustomKinveyPullCallback;
import com.kinvey.androidTest.callback.CustomKinveySyncCallback;
import com.kinvey.androidTest.callback.DefaultKinveyAggregateCallback;
import com.kinvey.androidTest.callback.DefaultKinveyClientCallback;
import com.kinvey.androidTest.callback.DefaultKinveyDeleteCallback;
import com.kinvey.androidTest.callback.DefaultKinveyReadCallback;
import com.kinvey.androidTest.callback.CustomKinveyPullCallback;
import com.kinvey.androidTest.callback.DefaultKinveyPushCallback;
import com.kinvey.androidTest.model.Person;
import com.kinvey.java.AbstractClient;
import com.kinvey.java.Query;
import com.kinvey.java.auth.KinveyAuthRequest;
import com.kinvey.java.core.KinveyCachedAggregateCallback;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.core.KinveyClientRequestInitializer;
import com.kinvey.java.model.AggregateType;
import com.kinvey.java.network.NetworkManager;
import com.kinvey.java.store.BaseDataStore;
import com.kinvey.java.store.StoreType;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

/**
 * Created by yuliya on 09/14/17.
 */

public class TestManager<T extends GenericJson> {

    public static final String TEST_USERNAME = "Test_UserName";
    public static final String TEST_USERNAME_2 = "Test_UserName_2";
    public static final String USERNAME = "test";
    public static final String USERNAME_USER = "test_user";
    public static final String USERNAME_FILE = "test_file";
    public static final String PASSWORD = "test";

    public void login(final String userName, final String password, final Client client) throws InterruptedException {
        if (!client.isUserLoggedIn()) {
            final CountDownLatch latch = new CountDownLatch(1);
            LooperThread looperThread;
            looperThread = new LooperThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        UserStore.login(userName, password, client, new KinveyClientCallback<User>() {
                            @Override
                            public void onSuccess(User result) {
                                assertNotNull(result);
                                latch.countDown();
                            }

                            @Override
                            public void onFailure(Throwable error) {
                                assertNull(error);
                                latch.countDown();
                            }
                        });
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            });
            looperThread.start();
            latch.await();
            looperThread.mHandler.sendMessage(new Message());
        }
    }

    public void logout(final Client client) throws InterruptedException {
        if (client.isUserLoggedIn()) {
            final CountDownLatch latch = new CountDownLatch(1);
            LooperThread looperThread;
            looperThread = new LooperThread(new Runnable() {
                @Override
                public void run() {
                    UserStore.logout(client, new KinveyClientCallback<Void>() {
                        @Override
                        public void onSuccess(Void result) {
                            latch.countDown();
                        }

                        @Override
                        public void onFailure(Throwable error) {
                            assertNull(error);
                            latch.countDown();
                        }
                    });
                }
            });
            looperThread.start();
            latch.await();
            looperThread.mHandler.sendMessage(new Message());
        }
    }

    public KinveyAuthRequest.Builder createBuilder(AbstractClient client) {
        String appKey = ((KinveyClientRequestInitializer) client.getKinveyRequestInitializer()).getAppKey();
        String appSecret = ((KinveyClientRequestInitializer) client.getKinveyRequestInitializer()).getAppSecret();
        return new KinveyAuthRequest.Builder(client.getRequestFactory().getTransport(),
                client.getJsonFactory(), client.getBaseUrl(), appKey, appSecret, null);
    }

    public DefaultKinveyClientCallback save(final DataStore<Person> store, final Person person) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyClientCallback callback = new DefaultKinveyClientCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.save(person, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyDeleteCallback delete(final DataStore<Person> store, final String id) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyDeleteCallback callback = new DefaultKinveyDeleteCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.delete(id, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyDeleteCallback delete(final DataStore<Person> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyDeleteCallback callback = new DefaultKinveyDeleteCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.delete(query, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyDeleteCallback deleteCustom(final DataStore<T> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyDeleteCallback callback = new DefaultKinveyDeleteCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.delete(query, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyDeleteCallback deleteCustom(final DataStore<T> store, final String id) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyDeleteCallback callback = new DefaultKinveyDeleteCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.delete(id, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyClientCallback<T> saveCustom(final DataStore<T> store, final T person) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyClientCallback<T> callback = new CustomKinveyClientCallback<T>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.save(person, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public void saveCustomInBackground(final DataStore<T> store, final T person) throws InterruptedException {
        final LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                try {
                    LooperThread.sleep(10000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                store.save(person, new KinveyClientCallback<T>() {
                    @Override
                    public void onSuccess(T result) {
                        System.out.println("saved " +result);
                    }

                    @Override
                    public void onFailure(Throwable error) {
                        System.out.println(error.getMessage());
                    }
                });
            }
        });
        looperThread.start();
    }

    public CustomKinveyListCallback<T> saveCustomList(final DataStore<T> store, final List<T> persons) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyListCallback<T> callback = new CustomKinveyListCallback<T>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.save(persons, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyReadCallback find(final DataStore<Person> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyReadCallback callback = new DefaultKinveyReadCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.find(query, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyClientCallback find(final DataStore<Person> store, final String id) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyClientCallback callback = new DefaultKinveyClientCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.find(id, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyReadCallback<T> findCustom(final DataStore<T> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyReadCallback<T> callback = new CustomKinveyReadCallback<T>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.find(query, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyPullCallback pullCustom(final DataStore<T> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyPullCallback callback = new CustomKinveyPullCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                if (query != null) {
                    store.pull(query, callback);
                } else {
                    store.pull(callback);
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyPullCallback pullCustom(final DataStore<T> store, final Query query, final int pageSize) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyPullCallback callback = new CustomKinveyPullCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                if (query != null) {
                    store.pull(query, pageSize, callback);
                } else {
                    store.pull(pageSize, callback);
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyPullCallback pullCustom(final DataStore<T> store, final Query query, final boolean isAutoPagination) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyPullCallback callback = new CustomKinveyPullCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                if (query != null) {
                    store.pull(query, isAutoPagination, callback);
                } else {
                    store.pull(isAutoPagination, callback);
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyPushCallback push(final DataStore<Person> store) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyPushCallback callback = new DefaultKinveyPushCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.push(callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public DefaultKinveyAggregateCallback calculation(final DataStore<T> store, final AggregateType aggregateType,
                                                      final ArrayList<String> fields, final String sumField, final Query query,
                                                      final KinveyCachedAggregateCallback cachedCallback) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyAggregateCallback callback = new DefaultKinveyAggregateCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.group(aggregateType, fields, sumField, query, callback, cachedCallback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveySyncCallback sync(final DataStore<T> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveySyncCallback callback = new CustomKinveySyncCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.sync(query, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveySyncCallback sync(final DataStore<T> store, final Query query, final boolean isAutoPagination) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveySyncCallback callback = new CustomKinveySyncCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.sync(query, isAutoPagination, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveySyncCallback sync(final DataStore<T> store, final Query query, final int pageSize) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveySyncCallback callback = new CustomKinveySyncCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                if (query != null) {
                    store.sync(query, pageSize, callback);
                } else {
                    store.sync(pageSize, callback);
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    //cleaning backend store (can be improved)
    @Deprecated
    public void cleanBackendDataStore(DataStore<Person> store) throws InterruptedException {
        DefaultKinveyDeleteCallback deleteCallback = delete(store, new Query().notEqual("age", "100500"));
        assertNull(deleteCallback.getError());
    }

    public void cleanBackend(DataStore<T> store, StoreType storeType) throws InterruptedException {
        if (storeType != StoreType.NETWORK) {
            sync(store, store.getClient().query());
        }
        DefaultKinveyDeleteCallback deleteCallback = deleteCustom(store, new Query().notEqual("age", "100500"));
        assertNull(deleteCallback.getError());
        if (storeType == StoreType.SYNC) {
            sync(store, store.getClient().query());
        }
    }

    public void createPersons(DataStore<Person> store, int n) throws IOException {
        for (int i = 0; i < n; i++) {
            Person person = new Person();
            person.setUsername(TEST_USERNAME + i);
            Person savedPerson = store.save(person);
            assertNotNull(savedPerson);
        }
    }

    public CustomKinveyClientCallback<Boolean> subscribe(final DataStore<T> store, final CustomKinveyLiveServiceCallback<T> liveServiceCallback) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyClientCallback<Boolean> callback = new CustomKinveyClientCallback<>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.subscribe(liveServiceCallback, callback);
            }
        });
        looperThread.start();
        latch.await(60, TimeUnit.SECONDS);
//        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyLiveServiceCallback subscribeSync(final DataStore<T> store) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyLiveServiceCallback<T> callback = new CustomKinveyLiveServiceCallback<T>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                try {
                    store.subscribe(callback);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public CustomKinveyLiveServiceCallback subscribeAsync(final DataStore<T> store) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final CustomKinveyLiveServiceCallback<T> callback = new CustomKinveyLiveServiceCallback<T>(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                try {
                    store.subscribe(callback);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    public BaseDataStore<T> mockBaseDataStore(AbstractClient client, String collectionName, Class className, StoreType storeType, NetworkManager networkManager) throws IllegalAccessException, InvocationTargetException, InstantiationException, NoSuchMethodException, ClassNotFoundException {
        Class<?> aClass = Class.forName("com.kinvey.java.store.BaseDataStore");
        Constructor<BaseDataStore> constructor = (Constructor<BaseDataStore>) aClass.getDeclaredConstructor(AbstractClient.class, String.class, Class.class, StoreType.class, NetworkManager.class);
        constructor.setAccessible(true);
        BaseDataStore baseDataStore = constructor.newInstance(client, collectionName, className, storeType, networkManager);
        return baseDataStore;
    }

    //use for Person.COLLECTION and for Person.class
    public long getCacheSize(StoreType storeType, Client client) {
        return client.getCacheManager().getCache(Person.COLLECTION, Person.class, storeType.ttl).get().size();
    }
}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/TestManager.kt
package com.kinvey.androidTest

import android.content.Context
import android.os.Looper
import android.os.Message
import androidx.test.platform.app.InstrumentationRegistry
import com.google.api.client.json.GenericJson
import com.kinvey.android.Client
import com.kinvey.android.model.User
import com.kinvey.android.store.DataStore
import com.kinvey.android.store.UserStore.Companion.login
import com.kinvey.android.store.UserStore.Companion.logout
import com.kinvey.androidTest.callback.*
import com.kinvey.androidTest.model.Person
import com.kinvey.androidTest.store.datastore.PaginationTest
import com.kinvey.java.AbstractClient
import com.kinvey.java.Query
import com.kinvey.java.auth.KinveyAuthRequest.Builder
import com.kinvey.java.core.KinveyCachedAggregateCallback
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.core.KinveyClientRequestInitializer
import com.kinvey.java.dto.BaseUser
import com.kinvey.java.model.AggregateType
import com.kinvey.java.network.NetworkManager
import com.kinvey.java.store.BaseDataStore
import com.kinvey.java.store.StoreType
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import java.io.IOException
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Created by yuliya on 09/14/17.
 */

class TestManager<T : GenericJson> {

    @Throws(InterruptedException::class, IOException::class)
    fun login(userName: String?, password: String?, client: Client<*>?) {
        val latch = CountDownLatch(1)
        if (client?.isUserLoggedIn == false) {
            Thread(Runnable {
                Looper.prepare()
                try {
                    login<User>(userName ?: "", password ?: "", client as AbstractClient<User>, object : KinveyClientCallback<User> {
                        override fun onSuccess(result: User?) {
                            assertNotNull(result)
                            latch.countDown()
                        }

                        override fun onFailure(error: Throwable?) {
                            assertNull(error)
                            latch.countDown()
                        }
                    })
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                Looper.loop()
            }).start()
        } else {
            latch.countDown()
        }
        latch.await()
    }

    @Throws(InterruptedException::class)
    fun logout(client: Client<*>?) {
        if (client?.isUserLoggedIn == true) {
            val latch = CountDownLatch(1)
            val looperThread: LooperThread
            looperThread = LooperThread(Runnable {
                logout(client as AbstractClient<BaseUser>, object : KinveyClientCallback<Void> {
                    override fun onSuccess(result: Void?) {
                        latch.countDown()
                    }

                    override fun onFailure(error: Throwable?) {
                        assertNull(error)
                        latch.countDown()
                    }
                })
            })
            looperThread.start()
            latch.await()
            looperThread.mHandler?.sendMessage(Message())
        }
    }

    fun createBuilder(client: AbstractClient<*>?): Builder<*> {
        val appKey = (client?.kinveyRequestInitializer as KinveyClientRequestInitializer?)?.appKey
        val appSecret = (client?.kinveyRequestInitializer as KinveyClientRequestInitializer?)?.appSecret
        return Builder<User>(client?.requestFactory?.transport,
                client?.jsonFactory!!, client?.baseUrl, appKey, appSecret, null)
    }

    @Throws(InterruptedException::class)
    fun save(store: DataStore<Person>?, person: Person?): DefaultKinveyClientCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyClientCallback(latch)
        val looperThread = LooperThread(Runnable { store?.save(person!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun delete(store: DataStore<Person>?, id: String?): DefaultKinveyDeleteCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyDeleteCallback(latch)
        val looperThread = LooperThread(Runnable { store?.delete(id, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun delete(store: DataStore<Person>?, query: Query?): DefaultKinveyDeleteCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyDeleteCallback(latch)
        val looperThread = LooperThread(Runnable { store?.delete(query!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> deleteCustom(store: DataStore<T>?, query: Query?): DefaultKinveyDeleteCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyDeleteCallback(latch)
        val looperThread = LooperThread(Runnable { store?.delete(query!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> deleteCustom(store: DataStore<T>?, id: String?): DefaultKinveyDeleteCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyDeleteCallback(latch)
        val looperThread = LooperThread(Runnable { store?.delete(id, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> saveCustom(store: DataStore<T>?, person: T?): CustomKinveyClientCallback<T> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyClientCallback<T>(latch)
        val looperThread = LooperThread(Runnable { store?.save(person!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> saveCustomInBackground(store: DataStore<T>?, person: T?) {
        val looperThread = LooperThread(Runnable {
            try {
                Thread.sleep(10000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            store?.save(person!!, object : KinveyClientCallback<T> {
                override fun onSuccess(result: T?) {
                    println("saved $result")
                }
                override fun onFailure(error: Throwable?) {
                    println(error?.message)
                }
            })
        })
        looperThread.start()
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> saveCustomList(store: DataStore<T>?, persons: List<T>?): CustomKinveyListCallback<T> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyListCallback<T>(latch)
        val looperThread = LooperThread(Runnable { store?.save(persons!!, callback as KinveyClientCallback<List<T>>) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun find(store: DataStore<Person>?, query: Query?): DefaultKinveyReadCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyReadCallback(latch)
        val looperThread = LooperThread(Runnable { store?.find(query!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun find(store: DataStore<Person>?, id: String?): DefaultKinveyClientCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyClientCallback(latch)
        val looperThread = LooperThread(Runnable { store?.find(id!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> findCustom(store: DataStore<T>?, query: Query?): CustomKinveyReadCallback<T> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyReadCallback<T>(latch)
        val looperThread = LooperThread(Runnable { store?.find(query!!, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> pullCustom(store: DataStore<T>?, query: Query?): CustomKinveyPullCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyPullCallback(latch)
        val looperThread = LooperThread(Runnable {
            if (query != null) {
                store?.pull(query, callback)
            } else {
                store?.pull(callback)
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> pullCustom(store: DataStore<T>?, query: Query?, pageSize: Int): CustomKinveyPullCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyPullCallback(latch)
        val looperThread = LooperThread(Runnable {
            if (query != null) {
                store?.pull(query, pageSize, callback)
            } else {
                store?.pull(pageSize, callback)
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> pullCustom(store: DataStore<T>?, query: Query?, isAutoPagination: Boolean): CustomKinveyPullCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyPullCallback(latch)
        val looperThread = LooperThread(Runnable {
            if (query != null) {
                store?.pull(query, isAutoPagination, callback)
            } else {
                store?.pull(isAutoPagination, callback)
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun push(store: DataStore<Person>?): DefaultKinveyPushCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyPushCallback(latch)
        val looperThread = LooperThread(Runnable { store?.push(callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> calculation(store: DataStore<T>?, aggregateType: AggregateType?,
                    fields: ArrayList<String>?, sumField: String?, query: Query?,
                    cachedCallback: KinveyCachedAggregateCallback?): DefaultKinveyAggregateCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyAggregateCallback(latch)
        val looperThread = LooperThread(Runnable { store?.group(aggregateType!!, fields!!, sumField, query!!, callback, cachedCallback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> sync(store: DataStore<T>?, query: Query?): CustomKinveySyncCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveySyncCallback(latch)
        val looperThread = LooperThread(Runnable { store?.sync(query, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> sync(store: DataStore<T>?, query: Query?, isAutoPagination: Boolean): CustomKinveySyncCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveySyncCallback(latch)
        val looperThread = LooperThread(Runnable { store?.sync(query, isAutoPagination, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> sync(store: DataStore<T>?, query: Query?, pageSize: Int): CustomKinveySyncCallback {
        val latch = CountDownLatch(1)
        val callback = CustomKinveySyncCallback(latch)
        val looperThread = LooperThread(Runnable {
            if (query != null) {
                store?.sync(query, pageSize, callback)
            } else {
                store?.sync(pageSize, callback)
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    //cleaning backend store (can be improved)
    @Deprecated("")
    @Throws(InterruptedException::class)
    fun cleanBackendDataStore(store: DataStore<Person>?) {
        val deleteCallback = delete(store, Query().notEqual("age", "100500"))
        assertNull(deleteCallback.error)
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> cleanBackend(store: DataStore<T>?, storeType: StoreType) {
        if (storeType !== StoreType.NETWORK) {
            sync(store, store?.client?.query()!!)
        }
        val deleteCallback = deleteCustom(store, Query().notEqual("age", "100500"))
        assertNull(deleteCallback.error)
        if (storeType === StoreType.SYNC) {
            sync(store, store?.client?.query()!!)
        }
    }

    @Throws(IOException::class)
    fun createPersons(store: DataStore<Person>?, n: Int) {
        for (i in 0 until n) {
            val person = Person()
            person.username = TEST_USERNAME + i
            val savedPerson = store?.save(person)
            assertNotNull(savedPerson)
        }
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> subscribe(store: DataStore<T>?, liveServiceCallback: CustomKinveyLiveServiceCallback<T>?): CustomKinveyClientCallback<Boolean> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyClientCallback<Boolean>(latch)
        val looperThread = LooperThread(Runnable { store?.subscribe(liveServiceCallback!!, callback) })
        looperThread.start()
        latch.await(60, TimeUnit.SECONDS)
//        looperThread.mHandler.sendMessage(new Message());


        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> subscribeSync(store: DataStore<T>?): CustomKinveyLiveServiceCallback<*> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyLiveServiceCallback<T>(latch)
        val looperThread = LooperThread(Runnable {
            try {
                store?.subscribe(callback)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    fun <T: GenericJson> subscribeAsync(store: DataStore<T>?): CustomKinveyLiveServiceCallback<*> {
        val latch = CountDownLatch(1)
        val callback = CustomKinveyLiveServiceCallback<T>(latch)
        val looperThread = LooperThread(Runnable {
            try {
                store?.subscribe(callback)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class, NoSuchMethodException::class, ClassNotFoundException::class)
    fun <T: GenericJson> mockBaseDataStore(client: AbstractClient<*>?, collectionName: String?, className: Class<*>?, storeType: StoreType?, networkManager: NetworkManager<T>?): BaseDataStore<T> {
        val aClass = Class.forName("com.kinvey.java.store.BaseDataStore")
        val constructor = aClass.getDeclaredConstructor(AbstractClient::class.java, String::class.java, Class::class.java, StoreType::class.java, NetworkManager::class.java) as Constructor<BaseDataStore<T>>
        constructor.isAccessible = true
        return constructor.newInstance(client, collectionName, className, storeType, networkManager)
    }

    //use for PaginationTest.COLLECTION and for Person.class
    fun getCacheSize(storeType: StoreType, client: Client<*>?): Long {
        return client?.cacheManager?.getCache(PaginationTest.COLLECTION, Person::class.java, storeType.ttl)?.get()?.size?.toLong() ?: 0L
    }

    companion object {
        const val TEST_USERNAME = "Test_UserName"
        const val TEST_USERNAME_2 = "Test_UserName_2"
        const val USERNAME = "testSecond"
        const val USERNAME_USER = "test_user"
        const val USERNAME_FILE = "test_file"
        const val PASSWORD = "testSecond"
    }
}